File: BinaryTrees/16/author/levelOrder.py
# Definition for a binary tree node.
from collections import deque


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def levelOrder(self, root: TreeNode) -> List[List[int]]:
        if not root:
            return []
        q = deque()
        q.append(root)
        ans = []
        li = []
        counter = 0
        numberToProcess = 1
        while len(q):
            node = q.popleft()
            li.append(node.val)
            counter += 1

            if node.left:
                q.append(node.left)
            if node.right:
                q.append(node.right)
            
            if counter == numberToProcess:
                ans.append(li)
                li = []
                counter = 0
                numberToProcess = len(q)
        return ans

File: BinaryTrees/16/author/levelOrder.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
var levelOrder = function (root) {
  if (!root) return [];
  let q = [root];
  numberToProcess = 1;
  const ans = [];
  let li = [];
  let counter = 0;
  while (q.length) {
    node = q.shift();
    li.push(node.val);
    counter++;
    
    if (node.left) q.push(node.left);
    if (node.right) q.push(node.right);

    if (counter === numberToProcess) {
      ans.push(li);
      li = [];
      counter = 0;
      numberToProcess = q.length;
    }
  }
  return ans;
};
